### DISCUSSION TITLE
Multiple Git Tokens Linux

### DISCUSSION THREAD

**Discussion by vmkoskin** [community/community#40657] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

Me and my coworker both have Personal Access Tokens created and have both successfully used them.  I have a shell script that clones a private git repo to a specific directory.  If I run the script it requires me to use my git login/PAT to do the clone. I tried to parameterize my PAT thinking we could each just pass it as a variable in the script but it doesn't work.  When I hard code my PAT it works fine.  I read about using git config with git credential helper but you can only configure one user it seems.  I would like to use HTTPs but I am not sure how we can either each pass our PAT when executing the script or how we can store them where if he logs in it grabs his vs mine.

**Comment by airtower-luna** [comment]
That shouldn't be a problem at all: `~/.gitconfig` or `~/.config/git/config` are per-user files, so each of you can configure a credential helper as you like.

That's unless you're sharing a user account, but you shouldn't be doing that anyway. :sweat_smile: 

As for the script, it should be able to use a token from a variable just like a hardcoded one, but without seeing the code it's impossible to say what's the problem. But as mentioned above you shouldn't need that. :slightly_smiling_face:

**Reply by vmkoskin** [reply]
Thank you for the reply! We log in as ourselves to the linux box but we have to sudo as another user to execute the script, would we still be able to use the git config this way?  Regarding the token I think it has some special characters because when I echo it appears to be correct but won't work when passing the variable.

**Reply by airtower-luna** [reply]
> We log in as ourselves to the linux box but we have to sudo as another user to execute the script, would we still be able to use the git config this way?

No, that won't work. The first question to consider here: Do you actually need to run the clone as that other user? Obviously that depends on what you do with the repository. For example if you install from it, you should be able to clone (or pull, or checkout) as your regular user, and run only the installation via sudo. I'd consider this the preferable approach, if at all possible. Minimum privilege and all that.

> Regarding the token I think it has some special characters because when I echo it appears to be correct but won't work when passing the variable.

The tokens I've seen didn't have any special characters. However, the mention of `sudo` most likely explains it: `sudo` cleans environment variables before executing the command, because many programs change their behavior based on environment variables, so keeping them while running as a different user (often `root`) could have all sorts of funny and dangerous effects. Take a look at its [`--preserve-env=list` option](https://manpages.debian.org/bullseye/sudo-ldap/sudo.8.en.html#preserve-env=list) if you want to try with variables again.
